### ISSUE TITLE
Formalize, document and version the internode protocol

### ISSUE THREAD

**Issue by andrewjstone** [andrewjstone/rabble#16]
Right now the code uses msgpack for internode serialization and the type sent between nodes is an [ExternalMsg](https://github.com/andrewjstone/rabble/blob/master/src/cluster/msg.rs#L22). In a long running system, changes to the internode messages will require versioning, and also some sort of version negotiation when nodes connect to each other. The explicit requirement is rolling upgrade of individual nodes in a long running cluster without taking the cluster down.

**Comment by andrewjstone** [comment]
Since the problem we are discussing is rolling upgrade of nodes, lets start by describing our
system, and it's current implementation.  We'll then go on to describe our needs and propose some
solutions, highlighting their benefits and drawbacks.

There are 2 nodes: `A` and `B`. Both nodes `A` and `B` communicate via a single message `Msg`.
`Msg`, in Rust looks like this:

```rust
#[derive(RustcEncodable, RustcDecodable)],
struct Msg {
    field1: u64
}
```

Currently, using msgpack and [rustc-serialize](https://doc.rust-lang.org/rustc-serialize/rustc_serialize/index.html), the implementation of the serialization and deserialization code for `Msg` is derived by the compiler. The huge benefit to this is that during initial development, creating new messages is easy, as you just have to define them as a normal rust type, and the code is automatically serializable. However, since we are constructing a distributed system, we must worry about more than easy creation of new messages. We must worry about forward and backwards compatibility for those messages. Let's say we want to add a second field to `Msg`, so that it looks like this:

```rust
#[derive(RustcEncodable, RustcDecodable)],
struct Msg {
    field1: u64,
    field2: u64
}
```

`Msg` is semantically the same, it's just adding some extra data in a new field. Now, since we can't
take down our running system all at once, we must upgrade one node at a time, which will result in
each node understanding a different version of `Msg`. Let's say we upgrade `A` and bring it back
online. `B` now serializes `Msg` into a msgpack type with 1 field (framed with a 4-byte size
header) and sends it over to `A`. `A` receives the (old) message and tries to deserialize it into a struct
with two fields. The deserialization fails, because it isn't backwards compatible: iit only knows the
current message expects 2 fields. Conversely, `A` could send a `Msg` to `B` which contains two
fields. This will probably actually succeed, (at least in this msgpack implementation), but will
silently ignore the second field. While this is probably what we want, the problem with this type of
forward compatibility is that it is **implicit**. Does the author of the new code expect that both
nodes handle the 2 field message the same?

In order to solve both backwards and forwards compatibility we essentially have 2 options:
 1. Version messages and negotiate versioning when nodes connect to each other
 2. Use an extensible protocol that allows us to explicitly add new fields to messages if they have
    the same semantic meaning.

For option 1. to work, when we upgrade node `A`, we must maintain both `MsgV1` and `MsgV2` and the
code to serialize and deserialize both. We can never just add a field. We always must introduce a new message. Due to the way rabble is architected (to send a single
statically typed message to processes) this means that if `A` is communicating with `B` it will need
to deserialize into `MsgV1` and then copy the data to a `MsgV2` struct so that it can send it to any
local processes. Conversely, for `A` to send `MsgV2` to `B`, it will need to create a new `MsgV1`
and serialize that. In reality, these messages are not small structs, but large enums, and this
introduces accidental complexity. Additionally, it creates unnecessary overhead when sending
messages across nodes, due to the required copying. Finally, we now have to add a version
negotiation protocol to the cluster servers to manage all this complexity. This seems like a serious
headache, and I'm ruling it out right now. There is a reason very few systems do things like this.

Option 2 is the way to go. Unknown fields are simply ignored in older versions of the code, while
deserializing old versions of messages to newer versions fills in a default or leaves the field
optional. Using option 2 however, now means that we have to explicitly define our messages outside
of simple rust types. Specifically, we need a schema and code generation to create the rust code to
manipulate our messages. Sticking with existing solutions, currently supported well in Rust, we end
up with essentially 2 options: [Cap'n Proto](https://capnproto.org/) or [Protocol Buffers](https://developers.google.com/protocol-buffers/). Fun fact: both Cap'n Proto and Protocol
buffers version 2 were written by Kenton Varda. Both solutions support unions which allow
us to define our single `Msg` as a rust `Enum` of many smaller messages. This allows evolution of
the protocol in a fairly complex manner, while directing subsets of the enum to processes that can
handle it. It's a fine compromise between static typing and dynamic typing, with all the performance
benefits of static typing (no boxing, inlining, etc..). It's also exactly how rabble currently
works.

`Cap'n Proto` is interesting because it uses `Zero-copy` parsing, which means that the same
structure manipulated in memory is what is sent over the wire. This means that there is no
serialization or deserialization overhead, although each object still needs to be allocated on the
heap on both creation and receipt like other serializers. However, to make use of this performance
benefit, all messages used by rabble must be represented as capnproto objects and manipulated using
the capnproto api.  This makes the rabble code less clear and also somewhat unergonomic to write. My
understanding is that the performance of accessing the capnproto object fields is comparable to raw
access of struct members, so this is not that much of a concern. Furthermore, with capnproto, since
all messages are allocated, we now must allocate for messages that are only sent locally, whereas
before we created them on the stack. (Note that sending over rust channels already does
allocations and copies the message into the channel, but using capnproto requires an additional
allocation).

Additionally, the capnproto in-memory format is extremely complicated, so much so that the pointer
layout algorithm is not described anywhere outside of the C++ code from what I can tell. In
contrast, there is [flatbuffers](https://google.github.io/flatbuffers/index.html) which essentially does the same thing as capnproto, but sacrifices direct memory access performance for a simpler vtable layout.  Flatbuffers is currently not under consideration due to the lack of a quality rust implementation. In addition to the complex in-memory layout, the implementation of capnproto uses a significant amount of unsafe code, although the quality of the code in general is high. There is also a lack of documentation, but that does not strike me as much of a problem after experimenting with the library.

A potential downside of the capnproto implementation in the context of rabble is that
deserialization does it's own allocation and framing. The current structure of rabble abstracts out
the serialization from the framing and allocating of the decode buffer. This complicates the code
and requires double allocation of the same space, or a re-write of how things work currently. Note
that the performance may not be that much of an issue, since protobuf also requires allocations
during decoding, and gains elsewhere would likely make up for this.

Finally, protobuf is widely used and well understood/supported, while capn'proto is backed by a
small company. Note however, that capnproto code is used in production and the format is stable at
this point.

Protobuf is a standard serialization protocol that allows unions and forward/backwards
compatibility. It's highly optimized and used inside google for most inter-service communication.
It's also already used in the client api of a currently closed sourced project that is the first
user of rabble. The big (supposed) downside of protobuf is that it requires serialization and
deserialization from the in memory object, thus imposing overhead when sending over the wire.
Protobuf also uses the current framing in rabble and is already supported.

Using protobuf in rabble would require a much smaller change to the code than switching to capnproto
for 2 reasons:

 1. Protobuf can use the current framing and is already supported by rabble
 2. Existing message types can remain in usage

In order to keep the existing message types, the generated types from the protobuf schema would be
copied into types of the current message format when deserialized. The opposite
operation is required for serialization: the rust types data must be copied into the protobuf
object. While neither of these operations requries allocations and the overhead from this is
minimal, they do require boring glue code. Ideally, we could write a macro to generate this code
for us. The one other benefit of doing things this way, is that if we ever made a newer version of rabble we could very easily swap out protobuf for a better serialization protocol if one comes along. If we use something like flatbuffers or capnproto, the changes are so huge that we are essentially stuck with them for life.

After having read through a bunch of capnproto code this weekend and trying to write a bit for rabble, I'm leaningheavily towards protobufs at this point.

Thanks to @justinsheehy for talking this through with me the other day.

**Comment by andrewjstone** [comment]
One more additional benefit of using protobufs is that it allows for serializing complex structures independent of the protobuf serialization. After decoding the protobuf wire format into the protobuf object we can then apply the extra deserialization for the complex structure (represented as bytes in the protobuf schema) and put the actual type back into the rust type.  For example, say we have a histogram that already has an efficient serialization mechanism. When the protobuf is decoded the field `histogram` will be a Vec<u8>. When constructing the rust type we can further decode the histogram into it's rust type and place it in the appropriate field. If using capnproto or flatbuffs we'd end up having it stored as bytes in the object. We'd either have to decode into a wrapper type or some thing magic. I suppose we could also use the same rust types we have now and copy from the capnproto object into them, but once again this defeats the main purpose of using capnproto and may actually be less efficient than the copying from protobuf as it isn't possible to move out large allocated objects like Vec's from capnproto objects. They *have* to be copied.

**Comment by andrewjstone** [comment]
After sleeping on this, it appears that we can get the same structural benefits of protobuf ( keeping our existing rust types) by copying from cap'n proto objects into the rust types on receipt, just as would be done with protobuf. This would not require allocations for local sends anymore, but would be more expensive than straight capn'proto use as the direct data type.

At this point it seems that a benchmark is in order comparing protobuf and capnproto where the objects are deserialized and copied into rust types. Since I already defined a complete schema in capnproto this should be fairly straightforward.

**Comment by andrewjstone** [comment]
I finally have some benchmarks that can shed some light on a direction to head in. Note that we always convert the deserialized type to a normal Rust type. Working with just the deserialized type, as Cap'n proto expects is **extremely** unergonomic and also prevents swapping the serialization format later.

```
running 9 tests
test serialize_correctly ... ignored
test bench_build_envelope_of_processes                     ... bench:       1,078 ns/iter (+/- 47)
test bench_build_pid                                       ... bench:         414 ns/iter (+/- 88)
test bench_deserialize_envelope_of_processes               ... bench:          33 ns/iter (+/- 4)
test bench_deserialize_pid                                 ... bench:          33 ns/iter (+/- 1)
test bench_deserialized_envelope_of_processes_to_rust_type ... bench:       1,728 ns/iter (+/- 691)
test bench_deserialized_pid_to_rust_type                   ... bench:         282 ns/iter (+/- 138)
test bench_serialize_envelope_of_processes                 ... bench:         179 ns/iter (+/- 54)
test bench_serialize_pid                                   ... bench:          51 ns/iter (+/- 16)

test result: ok. 0 passed; 0 failed; 1 ignored; 8 measured

     Running target/release/deps/protobuf-f21ebe2e6219340a

running 8 tests
test bench_build_envelope_of_processes                     ... bench:       1,117 ns/iter (+/- 216)
test bench_build_pid                                       ... bench:         164 ns/iter (+/- 24)
test bench_deserialize_envelope_of_processes               ... bench:       2,746 ns/iter (+/- 187)
test bench_deserialize_pid                                 ... bench:         403 ns/iter (+/- 161)
test bench_deserialize_pid_to_rust_type                    ... bench:          12 ns/iter (+/- 2)
test bench_deserialized_envelope_of_processes_to_rust_type ... bench:         156 ns/iter (+/- 29)
test bench_serialize_envelope_of_processes                 ... bench:         949 ns/iter (+/- 120)
test bench_serialize_pid                                   ... bench:         146 ns/iter (+/- 18)
```

Using the "envelope of processes", which is a realistically sized messages we have the following: 

Cap'n Proto total serialization time = 1257 ns
Cap'n Proto total deserialization time = 1761 ns

Protobuf total serialization time = 2066 ns
Protobuf total deserialization time = 2902 ns

It appears pretty clear that Cap'n proto is faster, but by less than a factor of 2. Furthermore, working with the protobuf API is much more ergonomic, which has some importance since users of rabble will have to define their own messages in that format. I also believe it's possible to remove some allocations by removing the use of oneof inside msg, which should further decrease the time taken to both encode and decode. I will need to benchmark this as well.

At this point I"m leaning pretty heavily toward using Protobuf. The only potential downside I see is the 1MB message limit, but for an actor system that is almost certainly a good thing. Any messages larger than that should be streamed via a higher level protocol.

**Comment by andrewjstone** [comment]
Further benchmark updates. It is essentially free to read from a rust struct (as expected), but it is not free to access a capnproto field. Additionally, even if we did use cap'n proto directly to save on deserialization, we'd pay this cost on every variable access in addition to the overhead of checking validity (which adds to the ergonomic nightmare).

```
test access_a_variable_from_reader                         ... bench:          41 ns/iter (+/- 16)
test access_a_variable_from_struct                         ... bench:           0 ns/iter (+/- 0)
```

**Comment by andrewjstone** [comment]
Surprisingly, removing oneof from the protobuf schema seemed to make things slower!

```
     Running target/release/deps/protobuf-f21ebe2e6219340a

running 8 tests
test bench_build_envelope_of_processes                     ... bench:       1,265 ns/iter (+/- 348)
test bench_build_pid                                       ... bench:         168 ns/iter (+/- 23)
test bench_deserialize_envelope_of_processes               ... bench:       2,997 ns/iter (+/- 394)
test bench_deserialize_pid                                 ... bench:         452 ns/iter (+/- 68)
test bench_deserialize_pid_to_rust_type                    ... bench:          12 ns/iter (+/- 6)
test bench_deserialized_envelope_of_processes_to_rust_type ... bench:         193 ns/iter (+/- 28)
test bench_serialize_envelope_of_processes                 ... bench:       1,036 ns/iter (+/- 193)
test bench_serialize_pid                                   ... bench:         165 ns/iter (+/- 25)

test result: ok. 0 passed; 0 failed; 0 ignored; 8 measured
```

**Comment by andrewjstone** [comment]
Actually, that may not be true, it may have just been cache warmup in prior runs. Variance is huge, and I'm seeing roughly equivalent numbers now. Either way, best to go with cleanest schema rather than trying to optimize via schema. Much bigger changes likely to come from changing pids and nodeids to be cached integers.

**Comment by andrewjstone** [comment]
Fixed by #17
